File: src/main/java/mffs/event/BlockPreMoveDelayedEvent.java
package mffs.event;

import mffs.DelayedEvent;
import mffs.IDelayedEventHandler;
import mffs.api.EventForceManipulate.EventPreForceManipulate;
import mffs.tile.TileForceManipulator;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraftforge.common.MinecraftForge;
import universalelectricity.api.vector.Vector3;
import universalelectricity.api.vector.VectorWorld;
import calclavia.lib.utility.MovementUtility;

/**
 * Removes the TileEntity
 * 
 * @author Calclavia
 * 
 */
public class BlockPreMoveDelayedEvent extends DelayedEvent
{
	private World world;
	private Vector3 position;
	private VectorWorld newPosition;

	public BlockPreMoveDelayedEvent(IDelayedEventHandler handler, int ticks, World world, Vector3 position, VectorWorld newPosition)
	{
		super(handler, ticks);
		this.world = world;
		this.position = position;
		this.newPosition = newPosition;
	}

	@Override
	protected void onEvent()
	{
		if (!this.world.isRemote)
		{
			// Do a final check before actually moving.
			if (((TileForceManipulator) this.handler).canMove(new VectorWorld(world, position), newPosition))
			{
				TileEntity tileEntity = this.position.getTileEntity(this.world);

				EventPreForceManipulate evt = new EventPreForceManipulate(this.world, this.position.intX(), this.position.intY(), this.position.intZ(), this.newPosition.intX(), this.newPosition.intY(), this.newPosition.intZ());
				MinecraftForge.EVENT_BUS.post(evt);

				if (!evt.isCanceled())
				{
					int blockID = this.position.getBlockID(this.world);
					int blockMetadata = this.position.getBlockMetadata(this.world);

					NBTTagCompound tileData = new NBTTagCompound();

					if (tileEntity != null)
					{
						tileEntity.writeToNBT(tileData);
					}

					MovementUtility.setBlockSneaky(this.world, this.position, 0, 0, null);
					this.handler.getQuedDelayedEvents().add(new BlockPostMoveDelayedEvent(this.handler, 0, this.world, this.position, this.newPosition, blockID, blockMetadata, tileEntity, tileData));
				}
			}
			else
			{
				((TileForceManipulator) this.handler).markFailMove = true;
			}
		}
	}
}

File: src/main/scala/mffs/mobilize/event/BlockPreMoveDelayedEvent.scala
package mffs.mobilize.event

import mffs.mobilize.TileForceMobilizer
import net.minecraft.init.Blocks
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.MinecraftForge
import resonant.api.mffs.EventForceManipulate
import resonant.lib.utility.MovementUtility
import universalelectricity.core.transform.vector.VectorWorld

/**
 * Removes the TileEntity
 *
 * @author Calclavia
 */
class BlockPreMoveDelayedEvent(_handler: IDelayedEventHandler, _ticks: Int, val startPosition: VectorWorld, val newPosition: VectorWorld) extends DelayedEvent(_handler, _ticks)
{
  protected override def onEvent
  {
    if (!startPosition.world.isRemote)
    {
      if ((handler.asInstanceOf[TileForceMobilizer]).canMove(startPosition, newPosition))
      {
        val tileEntity = startPosition.getTileEntity
        val evt = new EventForceManipulate.EventPreForceManipulate(startPosition.world, startPosition.xi, startPosition.yi, startPosition.zi, newPosition.xi, newPosition.yi, newPosition.zi)
        MinecraftForge.EVENT_BUS.post(evt)

        if (!evt.isCanceled)
        {
          val blockID = startPosition.getBlock
          val blockMetadata = startPosition.getBlockMetadata
          MovementUtility.setBlockSneaky(startPosition.world, this.startPosition, Blocks.air, 0, null)
          val tileData: NBTTagCompound = new NBTTagCompound
          if (tileEntity != null)
          {
            tileEntity.writeToNBT(tileData)
          }
          handler.queueEvent(new BlockPostMoveDelayedEvent(handler, 0, this.startPosition, this.newPosition, blockID, blockMetadata, tileEntity, tileData))
        }
      }
      else
      {
        (handler.asInstanceOf[TileForceMobilizer]).markFailMove = true
      }
    }
  }
}